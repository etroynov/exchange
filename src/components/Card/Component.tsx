import { FC } from 'react';
import getSymbolFromCurrency from 'currency-symbol-map';

import { Account } from '../../typings';

import styles from './index.module.css';

type Props = Account;

export const Card: FC<Props> = ({ title, type, balance }) => (
  <section className={styles.component} data-testid="card-container">
    <header className={styles.row}>
      <div className={styles.col}>
        <h2 className={styles.title} data-testid="card-title">{title}</h2>
        <h3 className={styles.type} data-testid="card-type">{type}</h3>
      </div>
    </header>

    <main className={styles.balance} data-testid="card-balance">
      {getSymbolFromCurrency(type)} {balance}
    </main>
  </section>
);
