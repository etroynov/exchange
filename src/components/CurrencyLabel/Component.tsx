import getSymbolFromCurrency from 'currency-symbol-map';

import styles from './index.module.css';

export const CurrencyLabel = ({ text = '0', type = '' }) => (
  <span className={styles.component} data-testid="currency-label">
    {text} {getSymbolFromCurrency(type)}
  </span>
);
