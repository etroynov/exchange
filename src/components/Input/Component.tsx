import { FC, ChangeEvent } from 'react';
import styles from './index.module.css';

type Props = {
  name: string;
  type?: string;
  value?: string | number;
  placeholder?: string;
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
};

export const Input: FC<Props> = ({ name, type = 'text', placeholder, value, onChange }) => {
  return (
    <input
      name={name}
      type={type}
      value={value}
      placeholder={placeholder}
      onChange={onChange}
      className={styles.component}
    />
  );
};
