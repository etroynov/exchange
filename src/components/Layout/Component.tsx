import { FC } from 'react';
import { Link } from 'react-router-dom';

import styles from './index.module.css';

import img from '../../logo.png';

export const Layout: FC = ({ children }) => (
  <div>
    <header className={styles.header}>
      <Link className={styles.logo} to="/">
        <img src={img} width="100" alt="" />
        <h1 className={styles.title}>lighthouse exchange</h1>
      </Link>
    </header>

    {children}
  </div>
);
