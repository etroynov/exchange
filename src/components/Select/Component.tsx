import { FC, ChangeEvent } from 'react';

import styles from './index.module.css';

type Props = {
  name: string;
  options: any[];
  selected?: string;
  onChange?: (e: ChangeEvent<HTMLSelectElement>) => void;
};

export const Select: FC<Props> = ({ name, selected, options, onChange }) => {
  return (
    <select
      value={selected}
      name={name}
      className={styles.component}
      onChange={onChange}
      data-testid="select"
    >
      {options.map(({ name, value }) => (
        <option key={name} value={value} data-testid="select-option">
          {name}
        </option>
      ))}
    </select>
  );
};
