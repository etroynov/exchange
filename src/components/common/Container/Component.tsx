import { FC } from 'react';
import styles from './index.module.css';

export const Container: FC = ({ children }) => (
  <div className={styles.component}>{children}</div>
);
