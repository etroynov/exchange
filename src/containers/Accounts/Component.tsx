import { useContext } from 'react';
import { Link } from 'react-router-dom';

import { Card } from '../../components';

import { AccountsContext } from '../../contexts';

import styles from './index.module.css';

export const AccountsContainer = () => {
  const { accounts } = useContext(AccountsContext);

  return (
    <section>
      <header className={styles.header}>
        <h2>All accounts</h2>
      </header>

      <main className={styles.list}>
        {accounts.map((account) => (
          <Link key={account.id} to={`/exchange/${account.type.toLowerCase()}`}>
            <Card {...account} />
          </Link>
        ))}
      </main>
    </section>
  );
};
