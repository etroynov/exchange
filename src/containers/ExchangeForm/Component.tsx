import getSymbolFromCurrency from 'currency-symbol-map';

import { useParams } from 'react-router-dom';
import { useContext, useState, useCallback, MouseEvent } from 'react';

import { Input, Select, CurrencyLabel } from '../../components';
import { AccountsContext, CurrenciesContext } from '../../contexts';

import type { Account } from '../../typings';

import styles from './index.module.css';

type Params = {
  type: string;
};

export const ExchangeFormContainer = () => {
  const { accounts, setAccounts } = useContext(AccountsContext);
  const { currencies } = useContext(CurrenciesContext);
  const { type } = useParams<Params>();

  const from = type.toUpperCase();
  const options = accounts.filter(
    (account) => account.type !== from
  ).map(
    ({ title, type, balance }) => ({
      name: `${title} | You have: ${balance.toFixed(2)} ${getSymbolFromCurrency(type)}`,
      value: type,
    })
  );

  const [fromAccount] = accounts.filter(
    (account) => account.type === type.toUpperCase()
  );
  const [toAccount] = accounts.filter(
    (account) => account.type !== type.toUpperCase()
  );

  const [currencyTo, setCurrencyTo] = useState(toAccount?.type);
  const [amountFrom, setAmountFrom] = useState('');
  const [amountTo, setAmountTo] = useState('');

  const handleCurrencyChange = useCallback(
    (e: MouseEvent) => {
      e.preventDefault();

      const _accounts = accounts.reduce<Account[]>((acc, curr) => {
        if (curr.type === fromAccount.type) {
          curr.balance -= Number(amountFrom);
        }

        if (curr.type === toAccount.type) {
          curr.balance += Number(amountTo);
        }

        acc.push(curr);

        return acc
      }, [])

      setAccounts(_accounts);
    },
    [accounts, amountFrom, amountTo, fromAccount.type, setAccounts, toAccount.type]
  );

  return (
    <section>
      <header className={styles.header}>
        <h2 className={styles.title}>
          Exchange:&nbsp;
          <CurrencyLabel text={amountFrom} type={from} />
          &nbsp;to&nbsp;
          <CurrencyLabel text={amountTo} type={currencyTo} />
        </h2>
      </header>

      <main>
        <form className={styles.form} action="">
          <fieldset className={styles.fieldset}>
            <div>
              <h3 className={styles.fromTitle}>{fromAccount?.title}</h3>
              <h4 className={styles.fromSubTitle}>
                You have: {fromAccount?.balance}&nbsp;
                {getSymbolFromCurrency(from)}
              </h4>
            </div>
            <Select
              name="to"
              selected={currencyTo}
              options={options}
              onChange={(e) => {
                setCurrencyTo(e.target.value);
              }}
            />
          </fieldset>

          <fieldset className={styles.fieldset}>
            <Input
              placeholder="You will spend"
              name="amountFrom"
              value={amountFrom}
              onChange={(e) => {
                const from = e.target.value;
                const to = (Number(from) * currencies[currencyTo]).toFixed(2);

                setAmountFrom(from);
                setAmountTo(to);
              }}
            />
            <Input
              placeholder="You will get"
              name="amountTo"
              value={amountTo}
              onChange={(e) => {
                const to = e.target.value;
                setAmountTo(to);
              }}
            />
          </fieldset>

          <div>
            <input
              type="submit"
              value="exchange"
              onClick={handleCurrencyChange}
            />
          </div>
        </form>
      </main>
    </section>
  );
};
