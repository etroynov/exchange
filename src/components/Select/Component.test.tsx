import { fireEvent, render } from '@testing-library/react';

import { Select } from './Component';

const options = [
  {
    name: 'USD - American Dollar 1',
    value: 25,
  },
  {
    name: 'USD - American Dollar 2',
    value: 30,
  },
  {
    name: 'USD - American Dollar 3',
    value: 35,
  },
];

describe('Card', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      expect(render(<Select name="test" options={options} />)).toMatchSnapshot();
    });
  });

  it('select should contain 3 option', () => {
    const { getAllByTestId } = render(<Select name="test" options={options} />);
    const els = getAllByTestId('select-option');

    expect(els.length).toBe(3);
  });

  it('select option should contain correct value', () => {
    const { getByTestId } = render(<Select name="test" options={options} />);
    const el = getByTestId('select');

    fireEvent.change(el, { target: { value: 30 }})

    expect(el).toHaveValue('30');
  });

  it('should unmount without errors', () => {
    const { unmount } = render(<Select name="test" options={options} />);

    expect(unmount).not.toThrowError();
  });
});
