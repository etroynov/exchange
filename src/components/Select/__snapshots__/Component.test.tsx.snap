// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Snapshots should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <select
        class="component"
        data-testid="select"
        name="test"
      >
        <option
          data-testid="select-option"
          value="25"
        >
          USD - American Dollar 1
        </option>
        <option
          data-testid="select-option"
          value="30"
        >
          USD - American Dollar 2
        </option>
        <option
          data-testid="select-option"
          value="35"
        >
          USD - American Dollar 3
        </option>
      </select>
    </div>
  </body>,
  "container": <div>
    <select
      class="component"
      data-testid="select"
      name="test"
    >
      <option
        data-testid="select-option"
        value="25"
      >
        USD - American Dollar 1
      </option>
      <option
        data-testid="select-option"
        value="30"
      >
        USD - American Dollar 2
      </option>
      <option
        data-testid="select-option"
        value="35"
      >
        USD - American Dollar 3
      </option>
    </select>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
